node/src/Pos/Block/Network/Retrieval.hs

```diff
@@ -125,12 +126,15 @@ retrievalWorkerImpl SendActions {..} =
                 let firstHeader = headers ^. _Wrapped . _neLast
                 in handleCHsValid enqueueMsg endedRecoveryVar nodeId
                                   firstHeader (headerHash header)
-        void $ enqueueMsg (MsgRequestBlockHeaders (Just (S.singleton nodeId))) $ \_ _ -> pure $ Conversation $ \conv ->
+        convs <- enqueueMsg (MsgRequestBlockHeaders (Just (S.singleton nodeId))) $ \_ _ -> pure $ Conversation $ \conv ->
             requestHeaders cont mgh nodeId conv
                 `finally` void (tryPutMVar endedRecoveryVar False)
-        -- Block until the conversation has ended.
-        whenM (readMVar endedRecoveryVar) $
-            logInfo "Recovery mode exited gracefully"
+        case M.null convs of
```

Comment:
It always gives a `Map NodeId (m t)` where `t` is the output of the conversation body.